packages/shared/lib/services/on-event-scripts.service.ts

```diff
@@ -84,20 +124,88 @@ export const onEventScriptService = {
+                    .select<(DBOnEventScript & { provider_config_key: string })[]>(['inserted.*', '_nango_configs.unique_key as provider_config_key'])
                     .from('inserted')
                     .join('_nango_configs', 'inserted.config_id', '_nango_configs.id');
-                return res;
+                return res.map(dbMapper.from);
             }
             return [];
         });
     },
-    getByConfig: async (configId: number, event: OnEventType): Promise<OnEventScript[]> => {
-        return db.knex.from<OnEventScript>(TABLE).where({ config_id: configId, active: true, event: toDbEvent(event) });
+
+    getByEnvironmentId: async (environmentId: number): Promise<OnEventScript[]> => {
+        const existingScriptsQuery = await db.knex
+            .select<(DBOnEventScript & { provider_config_key: string })[]>(`${TABLE}.*`, '_nango_configs.unique_key as provider_config_key')
+            .from(TABLE)
+            .join('_nango_configs', `${TABLE}.config_id`, '_nango_configs.id')
+            .where({
+                '_nango_configs.environment_id': environmentId,
+                [`${TABLE}.active`]: true
+            });
+        return existingScriptsQuery.map(dbMapper.from);
+    },
+
+    getByConfig: async (configId: number, event: OnEventType): Promise<DBOnEventScript[]> => {
+        return db.knex.from<DBOnEventScript>(TABLE).where({ config_id: configId, active: true, event: eventTypeMapper.toDb(event) });
+    },
+
+    diffChanges: async ({
+        environmentId,
+        onEventScriptsByProvider
+    }: {
+        environmentId: number;
+        onEventScriptsByProvider: OnEventScriptsByProvider[];
+    }): Promise<{
+        added: Omit<OnEventScript, 'id' | 'fileLocation' | 'createdAt' | 'updatedAt'>[];
+        deleted: OnEventScript[];
+        updated: OnEventScript[];
+    }> => {
+        const res: Awaited<ReturnType<typeof onEventScriptService.diffChanges>> = {
+            added: [],
+            deleted: [],
+            updated: []
+        };
+
+        const existingScripts = await onEventScriptService.getByEnvironmentId(environmentId);
+
+        // Create a map of existing scripts for easier lookup
+        const existingMap = new Map(existingScripts.map((script) => [`${script.configId}:${script.name}:${script.event}`, script]));
+
+        for (const provider of onEventScriptsByProvider) {
+            const config = await configService.getProviderConfig(provider.providerConfigKey, environmentId);
+            if (!config || !config.id) continue;
+
+            for (const script of provider.scripts) {
+                const key = `${config.id}:${script.name}:${script.event}`;
+
+                const maybeScript = existingMap.get(key);
+                if (maybeScript) {
+                    // Script already exists - it's an update
+                    res.updated.push(maybeScript);
+
+                    // Remove from map to track deletions
+                    existingMap.delete(key);
```

Comment:
could named `previousMap` (or something like this) for clarity, I didn't understood why we were removing from existing if it was indeed existing

Reply:
good idea. will fix